### PR TITLE
[meshcop] fix premature free of finalize message

### PR DESCRIPTION
**Pull request by wgtdkp**
This PR fix the bug that the `finalize message` will be prematurely freed if the joiner is started at the second time without waiting for the finish of the first initiation.

### CODE REVIEW

File: src/core/meshcop/joiner.cpp [hunk 1]
```diff
@@ -129,7 +129,10 @@ otError Joiner::Start(const char *     aPskd,
     if (error != OT_ERROR_NONE)
     {
         otLogInfoMeshCoP("Error %s while starting joiner", otThreadErrorToString(error));
-        FreeJoinerFinalizeMessage();
+        if (mState == OT_JOINER_STATE_IDLE)
```
Comment (librasungirl) [inline src/core/meshcop/joiner.cpp]: How about initializing `error` as `OT_ERROR_NONE` at the begining?

Line#101 has the status check already, here not necessary to check again if `error` properly initiazlied.

Comment (jwhui) [inline src/core/meshcop/joiner.cpp]: How about moving this check into `FreeJoinerFinalizeMessage()`? This change would also require moving the call to `FreeJoinerFinalizeMessage()` in `Joiner::Finish()` to after the `SetState(OT_JOINER_STATE_IDLE)` call. Thoughts?

Comment (wgtdkp) [reply src/core/meshcop/joiner.cpp]: Makes sense to me.
Done.

File: src/core/meshcop/joiner.cpp [hunk 2]
```diff
@@ -450,11 +450,16 @@ otError Joiner::PrepareJoinerFinalizeMessage(const char *aProvisioningUrl,
 
 void Joiner::FreeJoinerFinalizeMessage(void)
 {
+    VerifyOrExit(mState == OT_JOINER_STATE_IDLE);
```
Comment (jwhui) [inline src/core/meshcop/joiner.cpp]: nit: include the NULL check in this `VerifyOrExit`?

Comment (wgtdkp) [reply src/core/meshcop/joiner.cpp]: made it in separate `VerifyOrExit` to make it clearer.

**Pull request merged**